### PR TITLE
SECURITY: Require 1 uppercase, 1 lowercase, 1 digit, 1 symbol for passwords

### PR DESCRIPTION
**Pull request by sadleb**
### Description
Task: https://app.asana.com/0/1201131148207877/1201543544397148

Don't allow super insecure passwords going forward. This approach and the regex to check complexity is taken from [Devise documentation](https://github.com/heartcombo/devise/wiki/How-To:-Set-up-simple-password-complexity-requirements). Note that I enhanced the regex to allow more special characters.

Also, I setup a Honeycomb trigger to alert when there is possible brute force attack:
https://ui.honeycomb.io/braven/datasets/platform.braven.org/triggers/dsL8katLvfQ?view=/braven/datasets/platform.braven.org/triggers

Follow-up task for another attack vector that this PR does nothing to solve for: https://app.asana.com/0/1201131148207877/1201607387698620

### Motivation and Context
Without this change, it's likely that at some point someone's account will be compromised simply b/c they used a password like "password". We're not trying to enforce super secure passwords, we're not a bank, but they should have some bare level of complexity/strength so that a brute force attack would require enough guesses to trigger alerts.

We should do this now, after only 1 semester of using this new user database, in order to avoid a painful "everyone needs to reset their password" exercise down the road with many times more users in the database.

### Test plan
1. login with existing account just to make sure it doesn't impact existing credentials, only new ones.
1. Reset password flow. Try the following:
    1. too short
    1. mismatched
    1. not complex enough
    1. success
    1. login with new password
1. Register new account flow. Try the following:
    1. Make sure and run sync on Enrolled Participant first, make sure the User record is created with an empty password.
    1. too short
    1. mismatched
    1. not complex enough
    1. success
    1. logout
    1. log back in with password

### Screenshots (if appropriate):

Register Account (Desktop)
image

Register Account (Mobile)

Reset password (Desktop)
image

Reset password (Mobile)

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have added neccessary labels
- [ ] My PR title follows the convention in the guide
- [ ] I have assigned myself to the PR
- [ ] I have requested reviews from team members
- [ ] I have created a task for reviewers on PM tool
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by sadleb** [comment]
I've tested this thoroughly and want to get it live before the first batch of "create your account" emails go out on Jan 3rd, so I'm merging now but will ask for a retro-active code review when the other engineers are back in the office.

**Review by elenabaurkot — COMMENTED** [review]
I think I am having a little bit of a hard time figuring out where the password actually gets set and the complexity gets checked- Overall it looks good to me, but maybe we can look at that for a few minutes in a working session just to for understandings sake.

**Comment by sadleb** [comment]
> I think I am having a little bit of a hard time figuring out where the password actually gets set and the complexity gets checked- Overall it looks good to me, but maybe we can look at that for a few minutes in a working session just to for understandings sake.

So the `validate :password_complexity` part is Rails built in validation framework. Validation happens on `create`, `save`, or `update` of the model. See [here](https://guides.rubyonrails.org/active_record_validations.html#when-does-validation-happen-questionmark) for more info.

File: app/models/user.rb [hunk 1]
```diff
@@ -279,6 +280,27 @@ def password_required?
...
+  # calculating strength so a lot of my 8 char passwords would fail. We're just trying to
+  # prevent folks from using a password like "password" but we don't want to make it hard
+  # to pick a password that works. We have an alert for brute force attempts so any password
+  # that would require a brute force approach (as opposed to just trying a handful of really
+  # common passwords) should be fine to allow.
+  PASSWORD_COMPLEXITY_REGEX=/(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*~`()_+=<>|{}:;"',.#{Regexp.escape('\/[]')}])/
+  def password_complexity
+    return unless password_required?
+    return if password.blank? || password =~ PASSWORD_COMPLEXITY_REGEX
+
```
[the remark this answers is not in the document]
Comment (sadleb) [reply app/models/user.rb:300]: The `password.blank?` check isn't strictly required b/c we `return unless password_required?` and there is a separate validation in the Devise framework for minimum length. It's just there to be extra safe b/c what we really want to do is run `password =~ PASSWORD_COMPLEXITY_REGEX`. That part just says, "return true if the password matches the regex". Your summary of this method is correct though.

**Comment by elenabaurkot** [comment]
I understand that the validation occurs when those actions happen, I guess I was getting confused by the validation checking on password, which I can't find in the user schema, but I see encrypted_password is there so in my mind the validation would check on that, but I guess if it's already encrypted that doesn't really make sense. I just thought validations only occurred for things that were actually in the schema so I got a little lost.

**Review by rshipp — COMMENTED** [review]
requiring certain characters is explicitly against password strength best practices. some references here:

* https://en.wikipedia.org/wiki/Password_strength#Guidelines_for_strong_passwords
* https://spycloud.com/new-nist-guidelines/

**Comment by sadleb** [comment]
> requiring certain characters is explicitly against password strength best practices. some references here:
> 
> * https://en.wikipedia.org/wiki/Password_strength#Guidelines_for_strong_passwords
> * https://spycloud.com/new-nist-guidelines/

Interesting. I did not know that. Any recommendations on how to introduce some sort of basic complexity requirement without going crazy and doing a ton of work? See my note here about having tried to use the `strong_password` gem and abandoning that: https://github.com/bebraven/platform/pull/955/files#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR289-R295

**Comment by rshipp** [comment]
I would strongly recommend not implementing this ourselves. Barring that, the lowest hanging fruit is to enforce minimum length of 8 characters (which we already do), and deny common dictionary passwords (like "password").

**Comment by sadleb** [comment]
> I would strongly recommend not implementing this ourselves. Barring that, the lowest hanging fruit is to enforce minimum length of 8 characters (which we already do), and deny common dictionary passwords (like "password").

I'm thinking about just loading [this dictionary](https://github.com/danielmiessler/SecLists/blob/master/Passwords/Common-Credentials/10k-most-common.txt) of common passwords into memory and checking against that instead of using this regex. Good idea or bad in your opinion @rshipp ?

**Comment by rshipp** [comment]
Better than the alternatives, worse than 3rd party auth.

**Comment by sadleb** [comment]
> Better than the alternatives, worse than 3rd party auth.

Follow-up PR: https://github.com/bebraven/platform/pull/958

**Pull request merged**